graph/src/ipfs/cache.rs

```diff
@@ -0,0 +1,286 @@
+                let client = RedisClient::new(logger, path.as_ref())
+                    .await
+                    .map(Arc::new)
+                    .map_err(|e| IpfsError::InvalidCacheConfig {
+                        source: anyhow!("Failed to create IPFS Redis cache at {path}: {e}"),
+                    })?;
+                Ok(Cache::Redis { client })
+            }
+            Some(path) => {
+                let fs = LocalFileSystem::new_with_prefix(&path).map_err(|e| {
+                    IpfsError::InvalidCacheConfig {
+                        source: anyhow!(
+                            "Failed to create IPFS file based cache at {}: {}",
+                            path.display(),
+                            e
+                        ),
+                    }
+                })?;
+                debug!(logger, "Using IPFS file based cache"; "path" => path.display());
+                Ok(Cache::Disk {
+                    store: Arc::new(fs),
+                })
+            }
+            None => {
+                debug!(logger, "Using IPFS in-memory cache"; "capacity" => capacity, "max_entry_size" => max_entry_size);
+                Ok(Self::Memory {
+                    cache: Arc::new(Mutex::new(LruCache::with_capacity(capacity))),
+                    max_entry_size,
+                })
+            }
+        }
+    }
+
+    async fn find(&self, logger: &Logger, path: &ContentPath) -> Option<Bytes> {
+        match self {
+            Cache::Memory {
+                cache,
+                max_entry_size: _,
+            } => cache.lock().unwrap().get(path).cloned(),
+            Cache::Disk { store } => {
+                let log_err = |e: &object_store::Error| log_object_store_err(logger, e, false);
+
+                let path = Path::from(path.cid().to_string());
+                let object = store.get(&path).await.inspect_err(log_err).ok()?;
+                let data = object.bytes().await.inspect_err(log_err).ok()?;
+                Some(data)
+            }
+            Cache::Redis { client } => client
+                .get(path)
+                .await
+                .inspect_err(|e| log_redis_err(logger, e))
+                .ok()
+                .and_then(|data| if data.is_empty() { None } else { Some(data) }),
+        }
+    }
+
+    async fn insert(&self, logger: &Logger, path: ContentPath, data: Bytes) {
+        match self {
+            Cache::Memory { max_entry_size, .. } if data.len() > *max_entry_size => {
+                return;
+            }
+            Cache::Memory { cache, .. } => {
+                let mut cache = cache.lock().unwrap();
+
+                if !cache.contains_key(&path) {
+                    cache.insert(path.clone(), data.clone());
+                }
+            }
+            Cache::Disk { store } => {
+                let log_err = |e: &object_store::Error| log_object_store_err(logger, e, true);
+                let path = Path::from(path.cid().to_string());
```

Comment:
This is necessary to ensure that we cache the correct data

```suggestion
                let path = Path::from(path.to_string());
```